### PR TITLE
Create income endpoint

### PR DESCRIPTION
**Pull request by timsheremet**
Completed create income endpoint

- Tests
- Files
- Supporting files

One e2e test keeps failing. Pushing pr to collaborate on fixing the issue

### CODE REVIEW

File: packages/backend/e2e/createIncome.spec.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+          id: body.createdIncome.id,
+          ...fakeCreateIncomeInput,
+          createdAt: expect.any(String),
+          updatedAt: expect.any(String),
+        }),
+      )
+      const createdIncome = await db.query.incomeTable.findFirst({
+        where: (income, { eq }) => eq(income.id, body.createdIncome.id),
+      })
+      expect(createdIncome).toBeDefined()
```
Comment (nikitasheremet) [inline packages/backend/e2e/createIncome.spec.ts:58]: This isn't a necessary check since the next line impicitly checks that the object is defined

File: packages/backend/e2e/createIncome.spec.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      expect(createdIncome).toEqual(
+        expect.objectContaining({
```
Comment (nikitasheremet) [inline packages/backend/e2e/createIncome.spec.ts:60]: Here it would be enough to just compare the id of the returned created income. Since we don't need to test that the db query works correctly

File: packages/backend/e2e/createIncome.spec.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      })
+
+      expect(response.status()).toBe(STATUS_CREATED_201)
+      const body = await response.json()
+
+      expect(body).toHaveProperty('createdIncome')
+      expect(body.createdIncome).toHaveProperty('id')
+      expect(body.createdIncome).toEqual(
+        expect.objectContaining({
+          id: body.createdIncome.id,
```
Comment (nikitasheremet) [inline packages/backend/e2e/createIncome.spec.ts:49]: I would do `id: expect.any(String)` since this test kinda defeats the purpose since you are testing the assertion against itself in a way

File: packages/backend/e2e/createIncome.spec.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    test('should create a new income and return the income object', async ({ request }) => {
+      const response = await request.post(`${BASE_URL}/createincome`, {
+        data: fakeCreateIncomeInput,
+      })
+
+      expect(response.status()).toBe(STATUS_CREATED_201)
+      const body = await response.json()
+
+      expect(body).toHaveProperty('createdIncome')
+      expect(body.createdIncome).toHaveProperty('id')
```
Comment (nikitasheremet) [inline packages/backend/e2e/createIncome.spec.ts:46]: These two aren't necessary since the next assertion confirms this to be true

File: packages/backend/e2e/createIncome.spec.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  STATUS_CREATED_201,
+  STATUS_INTERNAL_SERVER_ERROR_500,
+  STATUS_UNPROCESSABLE_ENTITY_422,
+} from '../src/models/statusCodes'
+import crypto from 'crypto'
+import { DB_ERROR } from '../src/models/errors/repositoryErrors'
+import { connectToDb, db } from '../src/db'
+import { CreateIncomeInput } from '../src/income/validation/models'
+import { excludeFieldsAndAdd } from '../src/utilities/excludeFieldsAndAdd'
+import { id } from 'zod/v4/locales/index.cjs'
```
Comment (nikitasheremet) [inline packages/backend/e2e/createIncome.spec.ts:12]: What is this?

File: packages/backend/src/income/__tests__/handler/createIncomeHandler.test.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    it('should return 201 and expense', async () => {
+      // Arrange
+      mockService.mockResolvedValue({ id: 1, name: 'Lunch' })
+
+      // Act
+      await createIncomeHandler(fakeCtx)
+
+      // Assert
+      expect(fakeCtx.status).toBe(STATUS_CREATED_201)
+      expect(fakeCtx.body).toEqual({ createdIncome: { id: 1, name: 'Lunch' } })
```
Comment (nikitasheremet) [inline packages/backend/src/income/__tests__/handler/createIncomeHandler.test.ts:36]: nit: make `{id:1, name: Lunch}` into a variable and reuse it between the mockResolved and the assertion

File: packages/backend/src/income/__tests__/validation/createIncomeValidation.test.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    it('should throw a ValidationError for missing accountId', () => {
+      const input = { ...validInput, accountId: undefined }
+      expect(() => validateCreateIncomeInput(input)).toThrow(new ValidationError(VALIDATION_ERROR_ACCOUNTID_MISSING))
+    })
+    it('should throw a ValidationError for invalid accountId', () => {
+      const input = { ...validInput, accountId: 123 }
+      expect(() => validateCreateIncomeInput(input)).toThrow(new ValidationError(VALIDATION_ERROR_ACCOUNTID_TYPE))
+    })
+  })
+  describe('if name is missing, if name is a non-string or if name is empty', () => {
```
Comment (nikitasheremet) [inline packages/backend/src/income/__tests__/validation/createIncomeValidation.test.ts:48]: This needs additional unit tests for when name is not a string, and name is an empty string

File: packages/backend/src/income/__tests__/validation/createIncomeValidation.test.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        // Act & Assert
+        expect(() => validateCreateIncomeInput(input)).toThrow(new ValidationError(VALIDATION_ERROR_DATE_FORMAT))
+      })
+    })
+  })
+
+  describe('when input is valid', () => {
+    it('should not throw an error', () => {
+      const input = { ...validInput }
+      expect(() => validateCreateIncomeInput(input)).not.toThrow()
```
Comment (nikitasheremet) [inline packages/backend/src/income/__tests__/validation/createIncomeValidation.test.ts:122]: Just use validInput here no need to create a shallow copy

File: packages/backend/src/utilities/mappers/income/DBIncomeToDomainIncome.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { Income, IncomeDbRow } from '../../../models/income/income'
+
+export interface DbIncomeWithIncomeCategory extends IncomeDbRow {}
```
Comment (nikitasheremet) [inline packages/backend/src/utilities/mappers/income/DBIncomeToDomainIncome.ts:3]: This interface doesn't make sense here. Db Income doesn't have income category

File: packages/backend/src/server.ts [hunk 10]
```diff
@@ -21,6 +21,7 @@ import {
   deleteExpenseSubCategoryHandler,
 } from './expenseSubCategories/handler'
 import { connectToDb } from './db'
+import * as handler from './income/handler/createIncomeHandler'
```
Comment (nikitasheremet) [inline packages/backend/src/server.ts:24]: This should be `import {createIncomeHanlder} from './income/handler`

**Review by nikitasheremet — CHANGES_REQUESTED** [review]

File: packages/backend/e2e/createIncome.spec.ts (hunk 5 above)
Comment (timsheremet) [reply packages/backend/e2e/createIncome.spec.ts:12]: I don't even know, copilot might've added it randomly haha

File: packages/backend/src/income/__tests__/validation/createIncomeValidation.test.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    it('should throw a ValidationError for missing accountId', () => {
+      const input = { ...validInput, accountId: undefined }
+      expect(() => validateCreateIncomeInput(input)).toThrow(new ValidationError(VALIDATION_ERROR_ACCOUNTID_MISSING))
+    })
+    it('should throw a ValidationError for invalid accountId', () => {
+      const input = { ...validInput, accountId: 123 }
+      expect(() => validateCreateIncomeInput(input)).toThrow(new ValidationError(VALIDATION_ERROR_ACCOUNTID_TYPE))
+    })
+  })
+  describe('if name is missing, if name is a non-string or if name is empty', () => {
```
Comment (nikitasheremet) [inline packages/backend/src/income/__tests__/validation/createIncomeValidation.test.ts:50]: Nit: This should be "when name is invalid" to be consistent

**Review by nikitasheremet — CHANGES_REQUESTED** [review]
